### ISSUE TITLE
Graphical installer failure

### ISSUE THREAD

**Issue by josyoun** [fdopen/opam-repository-mingw#11]
I just tried the graphical installer, which unfortunately failed.  Not sure what the problem is, but here's the log:

```
Output folder: C:\OCaml64
Delete file: C:\Users\Joe\AppData\Local\Temp\nsd4736.tmp
Create folder: C:\Users\Joe\AppData\Local\Temp\nsd4736.tmp
Cygwin is being installed. Just follow the instructions!
Execute: C:\Users\Joe\AppData\Local\Temp\nsd4736.tmp\cygwin-dl.exe --quiet-mode --root C:\OCaml64 --no-admin --local-package-dir=C:\Users\Joe\AppData\Local\Temp\cygwin\ -g --wait --packages=bc,dash,diffutils,dos2unix,file,findutils,gawk,make,mintty,ncurses,patch,rlwrap,sed,tar,unzip,wget,which,xz,zsh,rsync,git,perl,m4,curl,wget,time,mingw64-x86_64-gcc-core >NUL 2>&1
Delete file: C:\Users\Joe\AppData\Local\Temp\nsd4736.tmp\cygwin-dl.exe
Remove folder: C:\Users\Joe\AppData\Local\Temp\nsd4736.tmp\
Testing cygwin ...
Copying skeleton files.
These files are for the users to personalise their cygwin experience.

They will never be overwritten nor automatically updated.

'./.bashrc' -> '/home/Joe//.bashrc'
'./.bash_profile' -> '/home/Joe//.bash_profile'
'./.inputrc' -> '/home/Joe//.inputrc'
'./.profile' -> '/home/Joe//.profile'
test
Output folder: C:\OCaml64\tmp
Extracting packages ...
Output folder: C:\OCaml64\tmp\OCaml64
Extract: install.sh
Extract: ocaml64.sh
Output folder: C:\OCaml64\tmp\OCaml64\bin
Extract: aspcud.exe
Extract: clasp.exe
Extract: cudf2lp.exe
Extract: flexlink.exe
Extract: gringo.exe
Extract: misc2012.lp
Extract: opam.exe
Extract: specification.lp
Output folder: C:\OCaml64\tmp\OCaml64\include
Extract: flexdll.h
Output folder: C:\OCaml64\tmp\OCaml64\lib
Output folder: C:\OCaml64\tmp\OCaml64\lib\flexdll
Extract: default.manifest
Extract: default_amd64.manifest
Extract: flexdll.h
Extract: flexdll_cygwin.o
Extract: flexdll_initer_cygwin.o
Extract: flexdll_initer_mingw.o
Extract: flexdll_initer_mingw64.o
Extract: flexdll_initer_msvc.obj
Extract: flexdll_initer_msvc64.obj
Extract: flexdll_mingw.o
Extract: flexdll_mingw64.o
Extract: flexdll_msvc.obj
Extract: flexdll_msvc64.obj
Extract: flexlink_mingw64test.exe
Output folder: C:\OCaml64\tmp\OCaml64\mkpasswd
Extract: mmkpasswd.exe
Extract: mmkpasswd.sh
Output folder: C:\OCaml64\tmp
Installing opam, flexdll and aspcud  ...
Creating a customized passwd file ...
Installing the OCaml compiler  ...
Please don't close this window until the repository is initialized
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg
    command on your system.
  - you won't be able to use darcs repositories unless you install the darcs
    command on your system.

[WARNING] No external solver found, one of aspucd, packup and mccs is
          recommended (see
          http://opam.ocaml.org/doc/Install.html#ExternalSolvers for details)
[WARNING] Recommended dependencies -- most packages rely on these:
            - cc


=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[mingw: git]
[mingw: git]
[mingw: git]
[mingw: git]
[mingw: git]
[mingw: git]
[ERROR] Could not synchronize mingw from
        "git+https://github.com/fdopen/opam-repository-mingw.git":
        "git fetch -q origin" exited with code 3328
[mingw] git+https://github.com/fdopen/opam-repository-mingw.git unavailable
Cannot find C:/OCaml64/home/Joe/.opam/compilers/4.02.3/4.02.3+mingw64c/4.02
[ERROR] Initialisation failed
OpamStd.OpamSys.Exit(1)
opam/OCaml installation failed
installation failed
```

For reference, I grabbed the 64-bit version from [here](https://fdopen.github.io/opam-repository-mingw/installation/), which has the date `20. Feb 2016`. In case it matters, I already had Cygwin64 installed on my system as well as MSYS.  I've not tried the manual installation, which may work, but I'm interested in this method because it'd be easier for me to set it up for others.

**Comment by fdopen** [comment]
Thanks for your report, it's detailed enough to investigate further. The cygwin and opam installation were successful, but the initial fetching of the github repo failed for an unknown reason (I'm unable to reproduce it and google doesn't give any hints for error code 3328). It's probably either a problem with cygwin's git or the network. Are you behind some kind of proxy? 

Open a shell of the already created cygwin installation (with either `C:\Ocaml64\cygwin.bat` or `C:\OCaml64\bin\mintty.exe -` and retry the failed command:

```bash
/usr/local/bin/opam init -y mingw 'https://github.com/fdopen/opam-repository-mingw.git' --comp 4.02.3+mingw64c --switch 4.02.3+mingw64c
```

If it fails again, the outputs of 
```bash
env
which git
```
could help.

And try to do something else with git:
```bash
cd /tmp
/usr/bin/git clone 'https://github.com/octocat/Spoon-Knife'
```
Does this command succeed?

I will look if I can tweak the installation script to output better diagnostic messages.

**Comment by josyoun** [comment]
Thanks for the quick reply.  Short answer is that it looks like git is silently failing with https, but works fine with git.  What's bizarre is that I've a couple of other git installations on this computer and they all work fine including some from cygwin.  As such, it appears that it's the git installed from this installer from cygwin that's having the trouble.  Here's some information:

```
/usr/local/bin/opam init -y mingw 'https://github.com/fdopen/opam-repository-mingw.git' --comp 4.02.3+mingw64c --switch 4.02.3+mingw64c
Checking for available remotes: rsync and local, git.
  - you won't be able to use mercurial repositories unless you install the hg
    command on your system.
  - you won't be able to use darcs repositories unless you install the darcs
    command on your system.

[WARNING] No external solver found, one of aspucd, packup and mccs is
          recommended (see
          http://opam.ocaml.org/doc/Install.html#ExternalSolvers for details)
[WARNING] Recommended dependencies -- most packages rely on these:
            - cc


=-=- Fetching repository information =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
[mingw: git]
[mingw: git]
[mingw: git]
[mingw: git]
[mingw: git]
[mingw: git]
[ERROR] Could not synchronize mingw from
        "git+https://github.com/fdopen/opam-repository-mingw.git":
        "git fetch -q origin" exited with code 3328
[mingw] git+https://github.com/fdopen/opam-repository-mingw.git unavailable
Cannot find C:/OCaml64/home/Joe/.opam/compilers/4.02.3/4.02.3+mingw64c/4.02.3+mingw64c.comp: 4.02.3+mingw64c is not a valid compiler name.
[ERROR] Initialisation failed
OpamStd.OpamSys.Exit(1)

```

 ```
$ env 
USERDOMAIN_ROAMINGPROFILE=DESKTOP
HOMEPATH=\Users\Joe
APPDATA=C:\Users\Joe\AppData\Roaming
ProgramW6432=C:\Program Files
HOSTNAME=DESKTOP
SHELL=/bin/bash
TERM=cygwin
PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 60 Stepping 1, GenuineIntel
PROFILEREAD=true
WINDIR=C:\Windows
PUBLIC=C:\Users\Public
OLDPWD=/home/Joe
ORIGINAL_PATH=/cygdrive/c/ProgramData/Oracle/Java/javapath:/cygdrive/c/Program Files/Python2.7:/cygdrive/c/Program Files/Python2.7/Scripts:/cygdrive/c/Windows/system32:/cygdrive/c/Windows:/cygdrive/c/Windows/System32/Wbem:/cygdrive/c/Windows/System32/WindowsPowerShell/v1.0:/cygdrive/c/Program Files/MATLAB/R2015a/bin:/cygdrive/c/Program Files (x86)/CMake/bin:/cygdrive/c/Software/flexdll:/cygdrive/c/Software/mingw-w64/x86_64-4.9.3-posix-seh-rt_v4-rev1/mingw64/bin
USERDOMAIN=DESKTOP-1K5H0GB
CommonProgramFiles(x86)=C:\Program Files (x86)\Common Files
OS=Windows_NT
ALLUSERSPROFILE=C:\ProgramData
USER=Joe
!::=::\
TEMP=/tmp
COMMONPROGRAMFILES=C:\Program Files\Common Files
USERNAME=Joe
PROCESSOR_LEVEL=6
ProgramFiles(x86)=C:\Program Files (x86)
PATH=/usr/local/bin:/usr/bin:/cygdrive/c/ProgramData/Oracle/Java/javapath:/cygdrive/c/Program Files/Python2.7:/cygdrive/c/Program Files/Python2.7/Scripts:/cygdrive/c/Windows/system32:/cygdrive/c/Windows:/cygdrive/c/Windows/System32/Wbem:/cygdrive/c/Windows/System32/WindowsPowerShell/v1.0:/cygdrive/c/Program Files/MATLAB/R2015a/bin:/cygdrive/c/Program Files (x86)/CMake/bin:/cygdrive/c/Software/flexdll:/cygdrive/c/Software/mingw-w64/x86_64-4.9.3-posix-seh-rt_v4-rev1/mingw64/bin
PSModulePath=C:\Windows\system32\WindowsPowerShell\v1.0\Modules\
PWD=/home/Joe
SYSTEMDRIVE=C:
LANG=en_US.UTF-8
USERPROFILE=C:\Users\Joe
TZ=America/Los_Angeles
PS1=\[\e]0;\w\a\]\n\[\e[32m\]\u@\h \[\e[33m\]\w\[\e[0m\]\n\$ 
LOGONSERVER=\\DESKTOP-1K5H0GB
CommonProgramW6432=C:\Program Files\Common Files
PROCESSOR_ARCHITECTURE=AMD64
LOCALAPPDATA=C:\Users\Joe\AppData\Local
!C:=C:\OCaml64\bin
ProgramData=C:\ProgramData
EXECIGNORE=*.dll
SHLVL=1
HOME=/home/Joe
PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC
HOMEDRIVE=C:
PROMPT=$P$G
COMSPEC=C:\Windows\system32\cmd.exe
TMP=/tmp
SYSTEMROOT=C:\Windows
PRINTER=Microsoft Print to PDF
PROCESSOR_REVISION=3c01
INFOPATH=/usr/local/info:/usr/share/info:/usr/info
PROGRAMFILES=C:\Program Files
NUMBER_OF_PROCESSORS=1
SESSIONNAME=Console
COMPUTERNAME=DESKTOP
_=/usr/bin/env
```

```
$ which git
/usr/bin/git

$ cygpath.exe -w /usr/bin/git
C:\OCaml64\bin\git
```

```
$ git --version
git version 2.7.0
```

```
$ /usr/bin/git clone 'https://github.com/octocat/Spoon-Knife'
Cloning into 'Spoon-Knife'...

$ /usr/bin/git clone 'git://github.com/octocat/Spoon-Knife'
Cloning into 'Spoon-Knife'...
remote: Counting objects: 16, done.
remote: Total 16 (delta 0), reused 0 (delta 0), pack-reused 16
Receiving objects: 100% (16/16), 2.67 KiB | 0 bytes/s, done.
Resolving deltas: 100% (2/2), done.
Checking connectivity... done
```

For reference, I'm not behind a proxy, but I am working through a virtual machine.

**Comment by fdopen** [comment]
Ok, thanks for your investigation. It's a problem with cygwin's git. I will update the installer scripts to detect such kinds of errors and give at least better error messages. But I'm not sure, what to do about it. (I couldn't find many similar reports via google, so it's hopefully an unusual error).

Could you also post the following output:
```
cygcheck /usr/libexec/git-core/git-remote-https.exe
```
Do you remember, if you did manually choose a mirror during the installation of cygwin? Or was everything done automatically?

It's a known problem that cygwin's setup.exe will reuse old options from previous installations, including the last used mirror. If this mirror is not in a proper state any longer or previously selected options don't make sense in the current context, the installation can fail. Cygwin's setup doesn't indicate it with a proper exit status code, if there were any errors. I have to check it manually and this is very error prone.

**Comment by josyoun** [comment]
As far as cycheck:
```

$ cygcheck.exe /usr/libexec/git-core/git-remote-https.exe
C:\OCaml64\usr\libexec\git-core\git-remote-https.exe
  C:\OCaml64\bin\cygcrypto-1.0.0.dll
    C:\OCaml64\bin\cygwin1.dll
      C:\Windows\system32\KERNEL32.dll
        C:\Windows\system32\ntdll.dll
        C:\Windows\system32\KERNELBASE.dll
    C:\OCaml64\bin\cygz.dll
  C:\OCaml64\bin\cygcurl-4.dll
    C:\OCaml64\bin\cyggssapi_krb5-2.dll
      C:\OCaml64\bin\cygk5crypto-3.dll
        C:\OCaml64\bin\cygkrb5support-0.dll
          C:\OCaml64\bin\cygintl-8.dll
            C:\OCaml64\bin\cygiconv-2.dll
      C:\OCaml64\bin\cygkrb5-3.dll
        C:\OCaml64\bin\cygcom_err-2.dll
          C:\OCaml64\bin\cyggcc_s-seh-1.dll
    C:\OCaml64\bin\cygidn-11.dll
    C:\OCaml64\bin\cyglber-2-4-2.dll
    C:\OCaml64\bin\cygldap-2-4-2.dll
      C:\OCaml64\bin\cygsasl2-3.dll
      C:\OCaml64\bin\cygssl-1.0.0.dll
    C:\OCaml64\bin\cygnghttp2-14.dll
    C:\OCaml64\bin\cygpsl-5.dll
      C:\OCaml64\bin\cygunistring-2.dll
    C:\OCaml64\bin\cygssh2-1.dll
```
As far as the mirror, previously, I'd manually chosen a mirror.  I had been on `cygwin.mirrors.hoobly.com`, but just switched to `cygwin.mirror.constant.com`.  I tried downgrading the git version installed by your package from 2.7.0 to 2.6.2 and I had the same error.  Then, I tried doing an upgrade on my other cygwin installation and now it turns out that that installation has a broken git as well.  I tried switching mirrors and reinstalling git and that didn't help either.  In case it helps to compare to your version:

```
$ md5sum `which git`
8074317b333068033fdfd6cbe1f4c80d */usr/bin/git

$ for file in `cygcheck.exe /usr/libexec/git-core/git-remote-https.exe`; do md5sum `cygpath $file | tr -d '\r'`; done
d75926e380fcf7ab24ea275110f8ccd5 */usr/libexec/git-core/git-remote-https.exe
d7088c9716e39a8259eadd5e7937d5f6 */usr/bin/cygcrypto-1.0.0.dll
9b39293de2e82a56e9e9cd9950c5e7bd */usr/bin/cygwin1.dll
038b10c8e735fe667da29b2e92a09b8a */cygdrive/c/Windows/system32/KERNEL32.dll
b3e7a635c248ebf3a9c630917bdd5fa0 */cygdrive/c/Windows/system32/ntdll.dll
e202b8613c3e9171d6d5c5c851719bb6 */cygdrive/c/Windows/system32/KERNELBASE.dll
aa103d6214e0ed401c8bdb486c110215 */usr/bin/cygz.dll
ddd9d9cd81207074eafc51514edc73b4 */usr/bin/cygcurl-4.dll
62b71efb2545bb6bc0eb31a7fbe0eb95 */usr/bin/cyggssapi_krb5-2.dll
d32a5d7779fd947a35619d9f088f293f */usr/bin/cygk5crypto-3.dll
3b0291a375bb45db3506e88d0f1350b8 */usr/bin/cygkrb5support-0.dll
4e3c76867c617f24bb28941e638055ce */usr/bin/cygintl-8.dll
13e186eb3050cfbc0a5642afecf66035 */usr/bin/cygiconv-2.dll
0ae16b7a23a7b920505b5bbe8b76ea98 */usr/bin/cygkrb5-3.dll
21a40a6dac454a0a1695cf84a0f1aaea */usr/bin/cygcom_err-2.dll
9346d7a1455a510f5370d8ad585e9481 */usr/bin/cyggcc_s-seh-1.dll
ae0e42483fc172702fc194edabc4601c */usr/bin/cygidn-11.dll
c698a9b51d938c3d8fa3353e6fef5704 */usr/bin/cyglber-2-4-2.dll
f614e350c77a1c18b7d3f20c8a4101c3 */usr/bin/cygldap-2-4-2.dll
437d58366f739fd878d1dc50e0fe4507 */usr/bin/cygsasl2-3.dll
1c066c6e7db9fd3f58fbfeddc4be6d76 */usr/bin/cygssl-1.0.0.dll
42cb08f264e5428f91fb796c58e1d9dd */usr/bin/cygnghttp2-14.dll
28a6063969dd4f4748f9b1ed3359d12c */usr/bin/cygpsl-5.dll
3a3cea5bba01daede19027e7aef6ddf9 */usr/bin/cygunistring-2.dll
593e678621a7f38bf3d390400e89b30b */usr/bin/cygssh2-1.dll
```

Thanks again for the help.

**Comment by fdopen** [comment]
Ok,
when I've tested my updated scripts to detect such kinds of error, I've got the same error. My mirror was just behind yesterday. The new openssl package is likely the culprit: https://cygwin.com/ml/cygwin/2016-03/msg00020.html 
If you manually downgrade openssl, curl, git, etc. should work again.

I will hopefully be fixed upstream, soon. Once it is fixed, I will test my updated scripts and upload an slightly changed installer that prints nicer error messages in such cases.
